File: app/Services/NotificationService.php
<?php

namespace App\Services;

use App\Models\Alert;
use App\Models\User;
use App\Jobs\SendEmailNotificationJob;
use App\Jobs\SendPushNotificationJob;
use App\Jobs\SendWhatsAppNotificationJob;
use Illuminate\Support\Facades\Log;

class NotificationService
{
    /**
     * Envoyer les notifications pour une alerte déclenchée
     */
    public function sendAlertNotifications(Alert $alert): void
    {
        $user = $alert->user;
        $product = $alert->product;

        if (!$user || !$product) {
            Log::warning("Cannot send notifications: Alert {$alert->id} missing user or product");
            return;
        }

        $updateData = [];
        $notificationsSent = false;

        // Email notification
        if ($user->email_notifications && !$alert->email_sent) {
            try {
                SendEmailNotificationJob::dispatch($alert);
                $updateData['email_sent'] = true;
                $notificationsSent = true;
            } catch (\Exception $e) {
                Log::error("Failed to queue email notification for alert {$alert->id}: " . $e->getMessage());
            }
        }

        // Push notification
        if ($user->push_notifications && !$alert->push_sent && $user->fcm_token) {
            try {
                SendPushNotificationJob::dispatch($alert);
                $updateData['push_sent'] = true;
                $notificationsSent = true;
            } catch (\Exception $e) {
                Log::error("Failed to queue push notification for alert {$alert->id}: " . $e->getMessage());
            }
        }

        // WhatsApp notification
        if ($user->whatsapp_notifications && !$alert->whatsapp_sent && $user->whatsapp_number) {
            try {
                SendWhatsAppNotificationJob::dispatch($alert);
                $updateData['whatsapp_sent'] = true;
                $notificationsSent = true;
            } catch (\Exception $e) {
                Log::error("Failed to queue WhatsApp notification for alert {$alert->id}: " . $e->getMessage());
            }
        }

        // Mettre à jour l'alerte si des notifications ont été envoyées
        if ($notificationsSent) {
            $alert->update($updateData);
        }
    }

    /**
     * Vérifier et déclencher les alerts pour un produit
     * 
     * @param mixed $product Le produit à vérifier
     * @param bool $sendNotifications Si true, envoie les notifications. Si false, met juste à jour les alerts.
     * @return array Liste des alerts déclenchés
     */
    public function checkAndTriggerAlerts($product, bool $sendNotifications = false): array
    {
        $triggeredAlerts = [];
        
        $alerts = \App\Models\Alert::where('product_id', $product->id)
            ->where('is_active', true)
            ->whereNull('triggered_at')
            ->with(['user', 'product'])
            ->get();

        foreach ($alerts as $alert) {
            $shouldTrigger = false;

            switch ($alert->alert_type) {
                case 'PRICE_DROP':
                    $shouldTrigger = $product->current_price <= $alert->target_price;
                    break;
                case 'PRICE_INCREASE':
                    $shouldTrigger = $product->current_price >= $alert->target_price;
                    break;
                case 'STOCK_AVAILABLE':
                    // Pour l'instant, on considère que le produit est toujours en stock
                    $shouldTrigger = true;
                    break;
            }

            if ($shouldTrigger) {
                $alert->update([
                    'triggered_at' => now(),
                ]);

                // Envoyer les notifications seulement si demandé
                if ($sendNotifications) {
                    $this->sendAlertNotifications($alert->fresh());
                }

                $triggeredAlerts[] = $alert;
            }
        }

        return $triggeredAlerts;
    }
}

File: services/notification_service.py
from models.alert import Alert
from models.product import Product
from datetime import datetime
import logging
from services.email_service import EmailService

logger = logging.getLogger(__name__)
email_service = EmailService()

class NotificationService:
    """Service for handling notifications and alerts"""
    
    def check_and_trigger_alerts(self, product: Product, send_notifications: bool = False) -> list:
        """Check and trigger alerts for a product"""
        triggered_alerts = []
        
        alerts = Alert.query.filter_by(
            product_id=product.id,
            is_active=True
        ).filter(Alert.triggered_at.is_(None)).all()
        
        for alert in alerts:
            should_trigger = False
            
            if alert.alert_type == 'PRICE_DROP':
                should_trigger = float(product.current_price) <= float(alert.target_price)
            elif alert.alert_type == 'PRICE_INCREASE':
                should_trigger = float(product.current_price) >= float(alert.target_price)
            elif alert.alert_type == 'STOCK_AVAILABLE':
                # For now, consider product always in stock
                should_trigger = True
            
            if should_trigger:
                alert.triggered_at = datetime.utcnow()
                alert.updated_at = datetime.utcnow()
                
                # Send notifications if requested
                if send_notifications:
                    self.send_alert_notifications(alert)
                
                triggered_alerts.append(alert)
        
        from extensions import db
        db.session.commit()
        
        return triggered_alerts
    
    def send_alert_notifications(self, alert: Alert):
        """Send notifications for a triggered alert"""
        user = alert.user
        product = alert.product
        
        if user.email_notifications and not alert.email_sent and user.email:
            try:
                success = email_service.send_price_alert_email(
                    user_email=user.email,
                    user_name=user.name or user.email.split('@')[0],
                    product_name=product.name,
                    product_url=product.url,
                    current_price=float(product.current_price),
                    target_price=float(alert.target_price),
                    alert_type=alert.alert_type
                )
                if success:
                    alert.email_sent = True
                    logger.info(f"Email alert sent to {user.email} for product {product.name}")
                else:
                    logger.error(f"Failed to send email alert to {user.email}")
            except Exception as e:
                logger.error(f"Error sending email alert: {str(e)}")
        
        if user.push_notifications and not alert.push_sent and user.fcm_token:
            # TODO: Send push notification
            alert.push_sent = True
        
        if user.whatsapp_notifications and not alert.whatsapp_sent and user.whatsapp_number:
            # TODO: Send WhatsApp notification
            alert.whatsapp_sent = True
        
        from extensions import db
        db.session.commit()

